ddl/worker.go

```diff
@@ -0,0 +1,275 @@
+// Copyright 2015 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ddl
+
+import (
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/ngaut/log"
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/meta"
+	"github.com/pingcap/tidb/model"
+)
+
+func (d *ddl) startJob(ctx context.Context, job *model.Job) error {
+	// for every DDL, we must commit current transaction.
+	if err := ctx.FinishTxn(false); err != nil {
+		return errors.Trace(err)
+	}
+
+	// alloc a global job id.
+	// add this job to job queue.
+	err := d.meta.RunInNewTxn(false, func(m *meta.TMeta) error {
+		var err error
+		job.ID, err = m.GenGlobalID()
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		err = m.EnQueueDDLJob(job)
+
+		return errors.Trace(err)
+	})
+
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// notice worker that we push a new job and wait the job done.
+	syncNotify(d.jobCh)
+
+	// check
+	jobID := job.ID
+	ticker := time.NewTicker(10 * time.Second)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-d.jobDoneCh:
+		case <-ticker.C:
+		}
+
+		job, err = d.getHistoryJob(jobID)
+		if err != nil {
+			log.Errorf("get history job err %v, check again", err)
+			continue
+		} else if job == nil {
+			log.Warnf("job %d is not in history, maybe not run", jobID)
+			continue
+		}
+
+		// if a job is a history table, the state must be JobDone or JobCancel.
+		if job.State == model.JobDone {
+			return nil
+		}
+
+		return errors.Errorf("job is %s, err :%v", job.State, job.Error)
+	}
+}
+
+func (d *ddl) getHistoryJob(id int64) (*model.Job, error) {
+	var job *model.Job
+
+	err := d.meta.RunInNewTxn(false, func(t *meta.TMeta) error {
+		var err1 error
+		job, err1 = t.GetHistoryDDLJob(id)
+		return errors.Trace(err1)
+	})
+
+	return job, errors.Trace(err)
+}
+
+func syncNotify(ch chan struct{}) {
+	select {
+	case ch <- struct{}{}:
```

Comment:
yes, if another server becomes the owner, after it does the job, it will send a notice too, but no place to receive this channel, so we must use the default brach to prevent blocking.